src/groupeddataframe/grouping.jl

```diff
@@ -122,50 +119,113 @@ _names(gd::GroupedDataFrame) = _names(gd.parent)
+wrap(x::AbstractMatrix) =
+    NamedTuple{Tuple(gennames(size(x, 2)))}(Tuple(x[:, i] for i in 1:size(x, 2)))
+wrap(x::Any) = (x1=x,)
+
+function do_call(f::Union{Function, Type}, gd::GroupedDataFrame,
+                 incols::Tuple, i::Integer)
+    idx = gd.idx[gd.starts[i]:gd.ends[i]]
+    f(map(c -> view(c, idx), incols)...)
+end
+function do_call(f::Union{Function, Type}, gd::GroupedDataFrame,
+                 incols::NamedTuple, i::Integer)
+    idx = gd.idx[gd.starts[i]:gd.ends[i]]
+    f(map(c -> view(c, idx), incols))
+end
+do_call(f::Union{Function, Type}, gd::GroupedDataFrame, incols::Nothing, i::Integer) =
+    f(gd[i])
+
+_nrow(df::AbstractDataFrame) = nrow(df)
+_nrow(x::NamedTuple{<:Any, <:Tuple{Vararg{AbstractVector}}}) =
+    isempty(x) ? 0 : length(x[1])
+_ncol(df::AbstractDataFrame) = ncol(df)
+_ncol(x::Union{NamedTuple, DataFrameRow}) = length(x)
 
 """
-    map(f::Union{Function, Type}, gd::GroupedDataFrame)
+    map(cols => f, gd::GroupedDataFrame)
+    map(f, gd::GroupedDataFrame)
 
 Apply a function to each group of rows and return a `GroupedDataFrame`.
 
-For each group in `gd`, `f` is passed a `SubDataFrame` view holding the corresponding rows.
+If the first argument is a `cols => f` pair, `cols` must be a column name or index, or
+a vector or tuple thereof, and `f` must be a callable. If `cols` is a single column index,
+`f` is callbed with a `SubArray` view into that column for each group; else, `f` is called
+with a named tuple holding `SubArray` views into these columns.
+
+If the first argument is a callable, it is passed a `SubDataFrame` view for each group,
+and the returned `DataFrame` then consists of the returned rows plus the grouping columns.
+Note that this second form is much slower than the first one due to type instability.
+
 `f` can return a single value, a row or multiple rows. The type of the returned value
 determines the shape of the resulting data frame:
 - A single value gives a data frame with a single column and one row per group.
-- A named tuple or a `DataFrameRow` gives a data frame with one column for each field
-  and one row per group.
+- A named tuple of single values or a `DataFrameRow` gives a data frame with one column
+  for each field and one row per group.
 - A vector gives a data frame with a single column and as many rows
   for each group as the length of the returned vector for that group.
-- A data frame or a matrix gives a data frame with the same columns and as many rows
-  for each group as the rows returned for that group.
+- A data frame, a named tuple of vectors or a matrix gives a data frame
+  with the same columns and as many rows for each group as the rows returned for that group.
 
 In all cases, the resulting `GroupedDataFrame` contains all the grouping columns in addition
-to those listed above. Note that `f` must always return the same type of object for
+to those listed above. Column names are automatically generated when necessary: for functions
```

Comment:
I'm not totally satisfied with using `x1` as the column name for functions taking several arguments. We could simply concatenate the names of all inputs with that of the function. In practice it's unlikely to get very long. But often functions taking multiple arguments will be anonymous functions, so suffix will be `_function`, which isn't very useful. At least that would be consistent. Thoughts?

Reply:
I think we can go for join of column names and function name although it is clumsy indeed.

I would assume that in this case anyway user will return a `NamedTuple` or `DataFrame` or something similar.

If someone wants to return only a vector probably one will use:
```
by(df, cols; colname = selector => function)
```
pattern as it is more convenient.

So the case when we have to generate a column name should be rare in practice (if we appropriately document the functionality to teach people the right patterns - which I intend to do anyway as we discussed on Discourse)